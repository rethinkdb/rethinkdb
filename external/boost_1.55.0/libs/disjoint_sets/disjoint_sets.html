<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>
<head>
  <meta http-equiv="Content-Language" content="en-us">
  <meta http-equiv="Content-Type" content="text/html; charset=us-ascii">

  <title>Boost Disjoint Sets</title>
</head>

<body bgcolor="#FFFFFF" link="#0000EE" text="#000000" vlink="#551A8B" alink=
"#FF0000">
  <img src="../../boost.png" alt="C++ Boost" width="277" height=
  "86"><br clear="none">

  <h1><a name="sec:disjoint-sets" id="sec:disjoint-sets"></a> Disjoint
  Sets</h1>
  <pre>
disjoint_sets&lt;Rank, Parent, FindCompress&gt;
</pre>

  <p>This is class that provides disjoint sets operations with <i>union by
  rank</i> and <i>path compression</i>. A disjoint-sets data structure
  maintains a collection <i>S = {S<sub>1</sub>, S<sub>2</sub>, ...,
  S<sub>k</sub>}</i> of disjoint sets. Each set is identified by a
  <i>representative</i> which is some member of of the set. Sets are
  represented by rooted trees which are encoded in the <tt>Parent</tt>
  property map. Two heuristics: "union by rank" and "path compression" are
  used to speed up the operations&nbsp;[<a href=
  "./bibliography.html#tarjan83:_data_struct_network_algo">1</a>, <a href=
  "./bibliography.html#clr90">2</a>].</p>

  <h3>Where Defined</h3><a href=
  "../../boost/pending/disjoint_sets.hpp"><tt>boost/pending/disjoint_sets.hpp</tt></a>

  <h3>Template Parameters</h3>

  <table border summary="">
    <tr>
      <td><tt>Rank</tt></td>

      <td>must be a model of <a href=
      "../property_map/doc/ReadWritePropertyMap.html">ReadWritePropertyMap</a>
      with an integer value type and a key type equal to the set's element
      type.</td>
    </tr>

    <tr>
      <td><tt>Parent</tt></td>

      <td>must be a model of <a href=
      "../property_map/doc/ReadWritePropertyMap.html">ReadWritePropertyMap</a>
      and the key and value type the same as the set's element type.</td>
    </tr>

    <tr>
      <td><tt>FindCompress</tt></td>

      <td>should be one of the find representative and path compress function
      objects.</td>
    </tr>
  </table>

  <h3>Example</h3>

  <p>A typical usage pattern for <tt>disjoint_sets</tt> can be seen in the
  <a href=
  "../graph/doc/kruskal_min_spanning_tree.html"><tt>kruskal_minimum_spanning_tree()</tt></a>
  algorithm. In this example, we call <tt>link()</tt> instead of
  <tt>union_set()</tt> because <tt>u</tt> and <tt>v</tt> were obtained from
  <tt>find_set()</tt> and therefore are already the representatives for their
  sets.</p>
  <pre>
  ...
  disjoint_sets&lt;Rank, Parent, FindCompress&gt; dsets(rank, p);
  
  for (ui  = vertices(G).first; ui != vertices(G).second; ++ui)
    dsets.make_set(*ui);
  ...
  while ( !Q.empty() ) {
    e = Q.front();
    Q.pop();
    u = dsets.find_set(source(e));
    v = dsets.find_set(target(e));
    if ( u != v ) {
      *out++ = e;
      dsets.link(u, v);
    }
  }
</pre>

  <h3>Members</h3>

  <table border summary="">
    <tr>
      <th>Member</th>

      <th>Description</th>
    </tr>

    <tr>
      <td><tt>disjoint_sets(Rank r, Parent p)</tt></td>

      <td>Constructor.</td>
    </tr>

    <tr>
      <td><tt>disjoint_sets(const disjoint_sets&amp; x)</tt></td>

      <td>Copy constructor.</td>
    </tr>

    <tr>
      <td><tt>template &lt;class Element&gt;<br>
      void make_set(Element x)</tt></td>

      <td>Creates a singleton set containing Element <tt>x</tt>.</td>
    </tr>

    <tr>
      <td><tt>template &lt;class Element&gt;<br>
      void link(Element x, Element y)</tt></td>

      <td>Union the two sets <i>represented</i> by element <tt>x</tt> and
      <tt>y</tt>.</td>
    </tr>

    <tr>
      <td><tt>template &lt;class Element&gt;<br>
      void union_set(Element x, Element y)</tt></td>

      <td>Union the two sets that <i>contain</i> elements <tt>x</tt> and
      <tt>y</tt>. This is equivalent to
      <tt>link(find_set(x),find_set(y))</tt>.</td>
    </tr>

    <tr>
      <td><tt>template &lt;class Element&gt;<br>
      Element find_set(Element x)</tt></td>

      <td>Return the representative for the set containing element
      <tt>x</tt>.</td>
    </tr>

    <tr>
      <td><tt>template &lt;class ElementIterator&gt;<br>
      std::size_t count_sets(ElementIterator first, ElementIterator
      last)</tt></td>

      <td>Returns the number of disjoint sets.</td>
    </tr>

    <tr>
      <td><tt>template &lt;class ElementIterator&gt;<br>
      void compress_sets(ElementIterator first, ElementIterator
      last)</tt></td>

      <td>Flatten the parents tree so that the parent of every element is its
      representative.</td>
    </tr>
  </table>

  <h3>Complexity</h3>

  <p>The time complexity is <i>O(m alpha(m,n))</i>, where <i>alpha</i> is the
  inverse Ackermann's function, <i>m</i> is the number of disjoint-set
  operations (<tt>make_set()</tt>, <tt>find_set()</tt>, and <tt>link()</tt>
  and <i>n</i> is the number of elements. The <i>alpha</i> function grows
  very slowly, much more slowly than the <i>log</i> function.</p>

  <h3>See Also</h3><a href=
  "../graph/doc/incremental_components.html"><tt>incremental_connected_components()</tt></a>
  <hr>
  <pre>
disjoint_sets_with_storage&lt;ID,InverseID,FindCompress&gt;
</pre>

  <p>This class manages the storage for the rank and parent properties
  internally. The storage is in arrays, which are indexed by element ID,
  hence the requirement for the <tt>ID</tt> and <tt>InverseID</tt> functors.
  The rank and parent properties are initialized during construction so the
  each element is in a set by itself (so it is not necessary to initialize
  objects of this class with the <a href=
  "../graph/doc/incremental_components.html#sec:initialize-incremental-components">
  <tt>initialize_incremental_components()</tt></a> function). This class is
  especially useful when computing the (dynamic) connected components of an
  <tt>edge_list</tt> graph which does not provide a place to store vertex
  properties.</p>

  <h3>Template Parameters</h3>

  <table border summary="">
    <tr>
      <th>Parameter</th>

      <th>Description</th>

      <th>Default</th>
    </tr>

    <tr>
      <td><tt>ID</tt></td>

      <td>must be a model of <a href=
      "../property_map/doc/ReadablePropertyMap.html">ReadablePropertyMap</a> that
      maps elements to integers between zero 0 and N, the total number of
      elements in the sets.</td>

      <td><tt>boost::identity_property_map</tt></td>
    </tr>

    <tr>
      <td><tt>InverseID</tt></td>

      <td>must be a model of <a href=
      "../property_map/doc/ReadablePropertyMap.html">ReadablePropertyMap</a> that
      maps integers to elements.</td>

      <td><tt>boost::identity_property_map</tt></td>
    </tr>

    <tr>
      <td><tt>FindCompress</tt></td>

      <td>should be one of the find representative and path compress function
      objects.</td>

      <td><tt>representative_with_full_path_compression</tt></td>
    </tr>
  </table>

  <h3>Members</h3>

  <p>This class has all of the members in <tt>disjoint_sets</tt> as well as
  the following members.</p>
  <pre>
disjoint_sets_with_storage(size_type n = 0,
                           ID id = ID(),
                           InverseID inv = InverseID())
</pre>Constructor.
  <pre>
template &lt;class ElementIterator&gt;
void disjoint_sets_with_storage::
  normalize_sets(ElementIterator first, ElementIterator last)
</pre>This rearranges the representatives such that the representative of
each set is the element with the smallest ID.<br>
  Postcondition: <tt>v &gt;= parent[v]</tt><br>
  Precondition: the disjoint sets structure must be compressed.<br>
  <hr>

  <h2><a name="sec:representative-with-path-halving" id=
  "sec:representative-with-path-halving"></a></h2>
  <pre>
representative_with_path_halving&lt;Parent&gt;
</pre>

  <p>This is a functor which finds the representative vertex for the same
  component as the element <tt>x</tt>. While traversing up the representative
  tree, the functor also applies the path halving technique to shorten the
  height of the tree.</p>
  <pre>
Element operator()(Parent p, Element x)
</pre>
  <hr>

  <h2><a name="sec:representative-with-full-path-compression" id=
  "sec:representative-with-full-path-compression"></a><br></h2>
  <pre>
representative_with_full_path_compression&lt;Parent&gt;
</pre>

  <p>This is a functor which finds the representative element for the set
  that element <tt>x</tt> belongs to.</p>
  <pre>
Element operator()(Parent p, Element x)
</pre>

  <p><br></p>
  <hr>

  <p><a href="http://validator.w3.org/check?uri=referer"><img border="0" src=
  "../../doc/images/valid-html401.png" alt="Valid HTML 4.01 Transitional"
  height="31" width="88"></a></p>

  <p>Revised 
  <!--webbot bot="Timestamp" s-type="EDITED" s-format="%d %B, %Y" startspan -->01
  December, 2006<!--webbot bot="Timestamp" endspan i-checksum="38508" --></p>

  <table summary="">
    <tr valign="top">
      <td nowrap><i>Copyright &copy; 2000</i></td>

      <td><i><a href="http://www.boost.org/people/jeremy_siek.htm">Jeremy Siek</a>, Univ.of
      Notre Dame (<a href="mailto:jsiek@lsc.nd.edu">jsiek@lsc.nd.edu</a>)<br>
      <a href="http://www.boost.org/people/liequan_lee.htm">Lie-Quan Lee</a>,
      Univ.of Notre Dame (<a href=
      "mailto:llee1@lsc.nd.edu">llee1@lsc.nd.edu</a>)<br>
      <a href="http://www.lsc.nd.edu/~lums">Andrew Lumsdaine</a>, Univ.of
      Notre Dame (<a href=
      "mailto:lums@lsc.nd.edu">lums@lsc.nd.edu</a>)</i></td>
    </tr>
  </table>

  <p><i>Distributed under the Boost Software License, Version 1.0. (See
  accompanying file <a href="../../LICENSE_1_0.txt">LICENSE_1_0.txt</a> or
  copy at <a href=
  "http://www.boost.org/LICENSE_1_0.txt">http://www.boost.org/LICENSE_1_0.txt</a>)</i></p>
</body>
</html>
