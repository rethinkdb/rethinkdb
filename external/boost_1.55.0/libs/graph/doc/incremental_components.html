<HTML>
<!--
     Copyright (c) Jeremy Siek, Lie-Quan Lee, and Andrew Lumsdaine 2000
    
     Distributed under the Boost Software License, Version 1.0.
     (See accompanying file LICENSE_1_0.txt or copy at
     http://www.boost.org/LICENSE_1_0.txt)
  -->
<Head>
<Title>Boost Graph Library: Incremental Connected Components</Title>
<style type="text/css">
  <!--
     .code
     {
       border-left-style: groove; 
       border-left-width: 1px; 
       padding-left: 2em; 
     }

  -->
</style>

<BODY BGCOLOR="#ffffff" LINK="#0000ee" TEXT="#000000" VLINK="#551a8b" 
        ALINK="#ff0000"> 
<IMG SRC="../../../boost.png" 
     ALT="C++ Boost" width="277" height="86"> 

<BR Clear>

<H1>Incremental Connected Components</H1>

<P>
This section describes a family of functions and classes that work
together to calculate the connected components of an undirected graph.
The algorithm used here is based on the disjoint-sets (fast
union-find) data structure&nbsp;[<A
HREF="bibliography.html#clr90">8</A>,<A
HREF="bibliography.html#tarjan83:_data_struct_network_algo">27</A>]
which is a good method to use for situations where the graph is
growing (edges are being added) and the connected components
information needs to be updated repeatedly. This method does not cover
the situation where edges are both added and removed from the graph,
hence it is called <b><i>incremental</i></b><a
href="bibliography.html#eppstein97:dynamic_graph">[42]</a> (and not
fully dynamic). The disjoint-sets class is described in Section <A
HREF="../../disjoint_sets/disjoint_sets.html">Disjoint Sets</A>.

<P>
The following five operations are the primary functions that you will
use to calculate and maintain the connected components.  The objects
used here are a graph <TT>g</TT>, a disjoint-sets structure <TT>ds</TT>,
and vertices <TT>u</TT> and <TT>v</TT>.

<P>

<UL>
<LI><TT>initialize_incremental_components(g, ds)</TT> 
<BR>
Basic initialization of the disjoint-sets structure. Each
    vertex in the graph <TT>g</TT> is in its own set.
</LI>
<LI><TT>incremental_components(g, ds)</TT> 
<BR>
The connected components are calculated based on the edges in the graph
    <TT>g</TT> and the information is embedded in <TT>ds</TT>.
</LI>
<LI><TT>ds.find_set(v)</TT> 
<BR>
Extracts the component information for vertex <TT>v</TT> from the
    disjoint-sets.
</LI>
<LI><TT>ds.union_set(u, v)</TT> 
<BR>
Update the disjoint-sets structure when edge <i>(u,v)</i> is added to the graph.
</LI>
</UL>

<P>

<H3>Complexity</H3>

<P>
The time complexity for the whole process is <i>O(V + E
alpha(E,V))</i> where <i>E</i> is the total number of edges in the
graph (by the end of the process) and <i>V</i> is the number of
vertices.  <i>alpha</i> is the inverse of Ackermann's function which
has explosive recursively exponential growth. Therefore its inverse
function grows <I>very</I> slowly. For all practical purposes
<i>alpha(m,n) <= 4</i> which means the time complexity is only
slightly larger than <i>O(V + E)</i>.

<P>

<H3>Example</H3>

<P>
Maintain the connected components of a graph while adding edges using
the disjoint-sets data structure. The full source code for this
example can be found in <a
href="../example/incremental_components.cpp"><TT>examples/incremental_components.cpp</TT></a>.

<P>
<PRE class="code">
using namespace boost;

int main(int argc, char* argv[]) 
{
  typedef adjacency_list <vecS, vecS, undirectedS> Graph;
  typedef graph_traits<Graph>::vertex_descriptor Vertex;
  typedef graph_traits<Graph>::vertices_size_type VertexIndex;

  const int VERTEX_COUNT = 6;
  Graph graph(VERTEX_COUNT);

  std::vector<VertexIndex> rank(num_vertices(graph));
  std::vector<Vertex> parent(num_vertices(graph));

  typedef VertexIndex* Rank;
  typedef Vertex* Parent;

  disjoint_sets<Rank, Parent> ds(&rank[0], &parent[0]);

  initialize_incremental_components(graph, ds);
  incremental_components(graph, ds);

  graph_traits<Graph>::edge_descriptor edge;
  bool flag;

  boost::tie(edge, flag) = add_edge(0, 1, graph);
  ds.union_set(0,1);

  boost::tie(edge, flag) = add_edge(1, 4, graph);
  ds.union_set(1,4);

  boost::tie(edge, flag) = add_edge(4, 0, graph);
  ds.union_set(4,0);

  boost::tie(edge, flag) = add_edge(2, 5, graph);
  ds.union_set(2,5);
    
  std::cout << "An undirected graph:" << std::endl;
  print_graph(graph, get(boost::vertex_index, graph));
  std::cout << std::endl;
    
  BOOST_FOREACH(Vertex current_vertex, vertices(graph)) {
    std::cout << "representative[" << current_vertex << "] = " <<
      ds.find_set(current_vertex) << std::endl;
  }

  std::cout << std::endl;

  typedef component_index<VertexIndex> Components;

  // NOTE: Because we're using vecS for the graph type, we're
  // effectively using identity_property_map for a vertex index map.
  // If we were to use listS instead, the index map would need to be
  // explicity passed to the component_index constructor.
  Components components(parent.begin(), parent.end());

  // Iterate through the component indices
  BOOST_FOREACH(VertexIndex current_index, components) {
    std::cout << "component " << current_index << " contains: ";

    // Iterate through the child vertex indices for [current_index]
    BOOST_FOREACH(VertexIndex child_index,
                  components[current_index]) {
      std::cout << child_index << " ";
    }

    std::cout << std::endl;
  }

  return (0);
}
</PRE>

<P>
<hr>
<p>

<H2><A NAME="sec:initialize-incremental-components"></A>
<TT>initialize_incremental_components</TT>
</H2>

<P>
<DIV ALIGN="left">
<TABLE CELLPADDING=3 border>
<TR><TH ALIGN="LEFT"><B>Graphs:</B></TH>
<TD ALIGN="LEFT">undirected</TD>
</TR>
<TR><TH ALIGN="LEFT"><B>Properties:</B></TH>
<TD ALIGN="LEFT">rank, parent (in disjoint-sets)</TD>
</TR>
<TR><TH ALIGN="LEFT"><B>Complexity:</B></TH>
<TD></TD>
</TR>
</TABLE>
</DIV>

<P>
<PRE>
template &lt;class VertexListGraph, class DisjointSets&gt; 
void initialize_incremental_components(VertexListGraph&amp; G, DisjointSets&amp; ds)
</PRE>

<P>
This prepares the disjoint-sets data structure for the incremental
connected components algorithm by making each vertex in the graph a
member of its own component (or set).

<P>

<H3>Where Defined</H3>

<P>
<a href="../../../boost/graph/incremental_components.hpp"><TT>boost/graph/incremental_components.hpp</TT></a>

<p>
<hr>
<P>

<H2><A NAME="sec:incremental-components"></A>
<TT>incremental_components</TT>
</H2>

<P>
<DIV ALIGN="left">
<TABLE CELLPADDING=3 border>
<TR><TH ALIGN="LEFT"><B>Graphs:</B></TH>
<TD ALIGN="LEFT">undirected</TD>
</TR>
<TR><TH ALIGN="LEFT"><B>Properties:</B></TH>
<TD ALIGN="LEFT">rank, parent (in disjoint-sets)</TD>
</TR>
<TR><TH ALIGN="LEFT"><B>Complexity:</B></TH>
<TD ALIGN="LEFT"><i>O(E)</i></TD>
</TR>
</TABLE>
</DIV>

<p>
<PRE>
template &lt;class EdgeListGraph, class DisjointSets&gt;
void incremental_components(EdgeListGraph&amp; g, DisjointSets&amp; ds)
</PRE>

<P>
This function calculates the connected components of the graph,
embedding the results in the disjoint-sets data structure.

<P>

<H3>Where Defined</H3>

<P>
<a href="../../../boost/graph/incremental_components.hpp"><TT>boost/graph/incremental_components.hpp</TT></a>

<P>

<H3>Requirements on Types</H3>

<P>

<UL>
<LI>The graph type must be a model of <a href="./EdgeListGraph.html">EdgeListGraph</a>.
</LI>
</UL>

<P>
<hr>
<p>

<H2><A NAME="sec:same-component">
<TT>same_component</TT></A>
</H2>

<P>
<DIV ALIGN="left">
<TABLE CELLPADDING=3 border>
<TR><TH ALIGN="LEFT"><B>Properties:</B></TH>
<TD ALIGN="LEFT">rank, parent (in disjoint-sets)</TD>
</TR>
<TR><TH ALIGN="LEFT"><B>Complexity:</B></TH>
<TD ALIGN="LEFT"><i>O(alpha(E,V))</i></TD>
</TR>
</TABLE>
</DIV>

<P>
<PRE>
template &lt;class Vertex, class DisjointSet&gt;
bool same_component(Vertex u, Vertex v, DisjointSet&amp; ds)
</PRE>

<P>
This function determines whether <TT>u</TT> and <TT>v</TT> are in the same
component.

<P>

<H3>Where Defined</H3>

<P>
<a href="../../../boost/graph/incremental_components.hpp"><TT>boost/graph/incremental_components.hpp</TT></a>

<P>

<H3>Requirements on Types</H3>

<P>

<UL>
<LI><TT>Vertex</TT> must be compatible with the rank and parent
    property maps of the <TT>DisjointSets</TT> data structure.
</LI>
</UL>

<P>
<hr>
<p>

<H2><A NAME="sec:component-index"></A>
<TT>component_index</TT>
</H2>

<p>
<PRE>
component_index&lt;Index&gt;
</PRE>

<P>
The <tt>component_index</tt> class provides an STL
container-like view for the components of the graph. Each component is
a container-like object, and access is provided via
the <TT>operator[]</TT>.  A <TT>component_index</TT> object is
initialized with the parents property in the disjoint-sets calculated
from the <TT>incremental_components()</TT> function.  Optionally, a
vertex -&gt; index property map is passed in
(<tt>identity_property_map</tt> is used by default).

<P>

<H3>Where Defined</H3>

<P>
<a href="../../../boost/graph/incremental_components.hpp"><TT>boost/graph/incremental_components.hpp</TT></a>

<P>

<H3>Members</H3>

<P>
 
<table border>

<tr>
<th>Member</th> <th>Description</th>
</tr>

<tr>
<td><tt>value_type/size_type</tt></td>
<td>
The type for a component index (same as <tt>Index</tt>).
</td>
</tr>

<tr>
<td><tt>size_type size()</tt></td>
<td>
Returns the number of components in the graph.
</td>
</tr>


<tr>
<td><tt>iterator/const_iterator</tt></td>
<td>
Iterators used to traverse the available component indices [0 to <tt>size()</tt>).
</td>
</tr>

<tr>
<td><tt>iterator begin() const</tt></td>
<td>
Returns an iterator at the start of the component indices (0).
</td>
</tr>

<tr>
<td><tt>iterator end() const</tt></td>
<td>
Returns an iterator past the end of the component indices (<tt>size()</tt>).
</td>
</tr>
 
<tr>
<td><tt>std::pair&lt;component_iterator, component_iterator&gt; operator[size_type index] const</tt></td>
<td>
Returns a pair of iterators for the component at <tt>index</tt> where <tt>index</tt> is in [0, <tt>size()</tt>).
</td>
</tr>

</table> 

<br>
<HR>
<TABLE>
<TR valign=top>
<TD nowrap>Copyright &copy; 2000-2001</TD><TD>
<A HREF="http://www.boost.org/people/jeremy_siek.htm">Jeremy Siek</A>,
Indiana University (<A
HREF="mailto:jsiek@osl.iu.edu">jsiek@osl.iu.edu</A>)<br>
<A HREF="http://www.boost.org/people/liequan_lee.htm">Lie-Quan Lee</A>, Indiana University (<A HREF="mailto:llee@cs.indiana.edu">llee@cs.indiana.edu</A>)<br>
<A HREF="http://www.osl.iu.edu/~lums">Andrew Lumsdaine</A>,
Indiana University (<A
HREF="mailto:lums@osl.iu.edu">lums@osl.iu.edu</A>)
</TD></TR></TABLE>

</BODY>
</HTML> 
