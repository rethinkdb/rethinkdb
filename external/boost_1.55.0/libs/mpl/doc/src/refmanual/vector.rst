.. Sequences/Classes//vector |10

vector
======

Description
-----------

``vector`` is a |variadic|, `random access`__, `extensible`__ sequence of types that 
supports constant-time insertion and removal of elements at both ends, and 
linear-time insertion and removal of elements in the middle. On compilers that 
support the ``typeof`` extension, ``vector`` is the simplest and in many cases the
most efficient sequence.

__ `Random Access Sequence`_
__ `Extensible Sequence`_

Header
------

+-------------------+-------------------------------------------------------+
| Sequence form     | Header                                                |
+===================+=======================================================+
| Variadic          | ``#include <boost/mpl/vector.hpp>``                   |
+-------------------+-------------------------------------------------------+
| Numbered          | ``#include <boost/mpl/vector/vector``\ *n*\ ``.hpp>`` |
+-------------------+-------------------------------------------------------+

Model of
--------

* |Variadic Sequence|
* |Random Access Sequence|
* |Extensible Sequence|
* |Back Extensible Sequence|
* |Front Extensible Sequence|


Expression semantics
--------------------

In the following table, ``v`` is an instance of ``vector``, ``pos`` and ``last`` are iterators 
into ``v``, ``r`` is a |Forward Sequence|, ``n`` is an |Integral Constant|, and ``x`` and 
|t1...tn| are arbitrary types.

+---------------------------------------+-----------------------------------------------------------+
| Expression                            | Semantics                                                 |
+=======================================+===========================================================+
| .. parsed-literal::                   | ``vector`` of elements |t1...tn|; see                     |
|                                       | |Variadic Sequence|.                                      |
|    vector<|t1...tn|>                  |                                                           |
|    vector\ *n*\ <|t1...tn|>           |                                                           |
+---------------------------------------+-----------------------------------------------------------+
| .. parsed-literal::                   | Identical to ``vector``\ *n*\ ``<``\ |t1...tn|\ ``>``;    |
|                                       | see |Variadic Sequence|.                                  |
|    vector<|t1...tn|>::type            |                                                           |
|    vector\ *n*\ <|t1...tn|>::type     |                                                           |
+---------------------------------------+-----------------------------------------------------------+
| ``begin<v>::type``                    | An iterator pointing to the beginning of ``v``;           |
|                                       | see |Random Access Sequence|.                             |
+---------------------------------------+-----------------------------------------------------------+
| ``end<v>::type``                      | An iterator pointing to the end of ``v``;                 |
|                                       | see |Random Access Sequence|.                             |
+---------------------------------------+-----------------------------------------------------------+
| ``size<v>::type``                     | The size of ``v``; see |Random Access Sequence|.          |
+---------------------------------------+-----------------------------------------------------------+
| ``empty<v>::type``                    | |true if and only if| the sequence is empty;              |
|                                       | see |Random Access Sequence|.                             |
+---------------------------------------+-----------------------------------------------------------+
| ``front<v>::type``                    | The first element in ``v``; see                           |
|                                       | |Random Access Sequence|.                                 |
+---------------------------------------+-----------------------------------------------------------+
| ``back<v>::type``                     | The last element in ``v``; see                            |
|                                       | |Random Access Sequence|.                                 |
+---------------------------------------+-----------------------------------------------------------+
| ``at<v,n>::type``                     | The ``n``\ th element from the beginning of ``v``; see    |
|                                       | |Random Access Sequence|.                                 |
+---------------------------------------+-----------------------------------------------------------+
| ``insert<v,pos,x>::type``             | A new ``vector`` of following elements:                   |
|                                       | [``begin<v>::type``, ``pos``), ``x``,                     |
|                                       | [``pos``, ``end<v>::type``); see |Extensible Sequence|.   |
+---------------------------------------+-----------------------------------------------------------+
| ``insert_range<v,pos,r>::type``       | A new ``vector`` of following elements:                   |
|                                       | [``begin<v>::type``, ``pos``),                            |
|                                       | [``begin<r>::type``, ``end<r>::type``)                    |
|                                       | [``pos``, ``end<v>::type``); see |Extensible Sequence|.   |
+---------------------------------------+-----------------------------------------------------------+
| ``erase<v,pos>::type``                | A new ``vector`` of following elements:                   |
|                                       | [``begin<v>::type``, ``pos``),                            |
|                                       | [``next<pos>::type``, ``end<v>::type``); see              |
|                                       | |Extensible Sequence|.                                    |
+---------------------------------------+-----------------------------------------------------------+
| ``erase<v,pos,last>::type``           | A new ``vector`` of following elements:                   |
|                                       | [``begin<v>::type``, ``pos``),                            |
|                                       | [``last``, ``end<v>::type``); see |Extensible Sequence|.  |
+---------------------------------------+-----------------------------------------------------------+
| ``clear<v>::type``                    | An empty ``vector``; see |Extensible Sequence|.           |
+---------------------------------------+-----------------------------------------------------------+
| ``push_back<v,x>::type``              | A new ``vector`` of following elements:                   | 
|                                       | |begin/end<v>|, ``x``;                                    |
|                                       | see |Back Extensible Sequence|.                           |
+---------------------------------------+-----------------------------------------------------------+
| ``pop_back<v>::type``                 | A new ``vector`` of following elements:                   |
|                                       | [``begin<v>::type``, ``prior< end<v>::type >::type``);    |
|                                       | see |Back Extensible Sequence|.                           |
+---------------------------------------+-----------------------------------------------------------+
| ``push_front<v,x>::type``             | A new ``vector`` of following elements:                   |
|                                       | ``x``, |begin/end<v>|; see |Front Extensible Sequence|.   |
+---------------------------------------+-----------------------------------------------------------+
| ``pop_front<v>::type``                | A new ``vector`` of following elements:                   |
|                                       | [``next< begin<v>::type >::type``, ``end<v>::type``);     |
|                                       | see |Front Extensible Sequence|.                          |
+---------------------------------------+-----------------------------------------------------------+


Example
-------

.. parsed-literal::
    
    typedef vector<float,double,long double> floats;
    typedef push_back<floats,int>::type types;

    BOOST_MPL_ASSERT(( |is_same|\< at_c<types,3>::type, int > ));


See also
--------

|Sequences|, |Variadic Sequence|, |Random Access Sequence|, |Extensible Sequence|, |vector_c|, |list|


.. copyright:: Copyright ©  2001-2009 Aleksey Gurtovoy and David Abrahams
   Distributed under the Boost Software License, Version 1.0. (See accompanying
   file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
