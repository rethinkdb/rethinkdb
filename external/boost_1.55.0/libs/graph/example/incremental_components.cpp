//=======================================================================
// Copyright 1997, 1998, 1999, 2000 University of Notre Dame.
// Copyright 2009 Trustees of Indiana University.
// Authors: Andrew Lumsdaine, Lie-Quan Lee, Jeremy G. Siek, Michael Hansen
//
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
//=======================================================================
#include <iostream>
#include <vector>

#include <boost/foreach.hpp>
#include <boost/graph/adjacency_list.hpp>
#include <boost/graph/graph_utility.hpp>
#include <boost/graph/incremental_components.hpp>
#include <boost/pending/disjoint_sets.hpp>

/*

  This example shows how to use the disjoint set data structure
  to compute the connected components of an undirected, changing
  graph.

  Sample output:

  An undirected graph:
  0 <--> 1 4 
  1 <--> 0 4 
  2 <--> 5 
  3 <--> 
  4 <--> 1 0 
  5 <--> 2 

  representative[0] = 1
  representative[1] = 1
  representative[2] = 5
  representative[3] = 3
  representative[4] = 1
  representative[5] = 5

  component 0 contains: 4 1 0 
  component 1 contains: 3 
  component 2 contains: 5 2 

 */

using namespace boost;

int main(int argc, char* argv[]) 
{
  typedef adjacency_list <vecS, vecS, undirectedS> Graph;
  typedef graph_traits<Graph>::vertex_descriptor Vertex;
  typedef graph_traits<Graph>::vertices_size_type VertexIndex;

  const int VERTEX_COUNT = 6;
  Graph graph(VERTEX_COUNT);

  std::vector<VertexIndex> rank(num_vertices(graph));
  std::vector<Vertex> parent(num_vertices(graph));

  typedef VertexIndex* Rank;
  typedef Vertex* Parent;

  disjoint_sets<Rank, Parent> ds(&rank[0], &parent[0]);

  initialize_incremental_components(graph, ds);
  incremental_components(graph, ds);

  graph_traits<Graph>::edge_descriptor edge;
  bool flag;

  boost::tie(edge, flag) = add_edge(0, 1, graph);
  ds.union_set(0,1);

  boost::tie(edge, flag) = add_edge(1, 4, graph);
  ds.union_set(1,4);

  boost::tie(edge, flag) = add_edge(4, 0, graph);
  ds.union_set(4,0);

  boost::tie(edge, flag) = add_edge(2, 5, graph);
  ds.union_set(2,5);
    
  std::cout << "An undirected graph:" << std::endl;
  print_graph(graph, get(boost::vertex_index, graph));
  std::cout << std::endl;
    
  BOOST_FOREACH(Vertex current_vertex, vertices(graph)) {
    std::cout << "representative[" << current_vertex << "] = " <<
      ds.find_set(current_vertex) << std::endl;
  }

  std::cout << std::endl;

  typedef component_index<VertexIndex> Components;

  // NOTE: Because we're using vecS for the graph type, we're
  // effectively using identity_property_map for a vertex index map.
  // If we were to use listS instead, the index map would need to be
  // explicitly passed to the component_index constructor.
  Components components(parent.begin(), parent.end());

  // Iterate through the component indices
  BOOST_FOREACH(VertexIndex current_index, components) {
    std::cout << "component " << current_index << " contains: ";

    // Iterate through the child vertex indices for [current_index]
    BOOST_FOREACH(VertexIndex child_index,
                  components[current_index]) {
      std::cout << child_index << " ";
    }

    std::cout << std::endl;
  }

  return (0);
}

