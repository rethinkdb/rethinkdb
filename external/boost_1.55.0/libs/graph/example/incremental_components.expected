An undirected graph:
0 <--> 1 4 
1 <--> 0 4 
2 <--> 5 
3 <--> 
4 <--> 1 0 
5 <--> 2 

representative[0] = 1
representative[1] = 1
representative[2] = 5
representative[3] = 3
representative[4] = 1
representative[5] = 5

component 0 contains: 1 4 0 
component 1 contains: 3 
component 2 contains: 5 2 
